### PR TITLE
Ez leka/build optimizations

### PR DESCRIPTION
**Pull request by ez-leka**
Local build no longer needs /opt/kontain. Everything is build into build/opt/kontain directory. 

To build for testing locally use:
make -j 
make test

To build for testing with docker, you have to build with docker
make -j withdocker
make testing-image
make test-with docker

When building for release outside of release.yaml use 
make -j RPATH-/opt/kontain

### CODE REVIEW

File: tests/gdb_forker_test.c [hunk 1]
```diff
@@ -47,10 +48,10 @@ int main(int argc, char* argv[])
       char* new_argv[2];
       new_argv[0] = payload;
       new_argv[1] = NULL;
-      char* new_envp[1];
-      new_envp[0] = NULL;
+      // char* new_envp[1];
```
Comment (sv641) [inline tests/gdb_forker_test.c:51]: remove unused

**Review by sv641 — APPROVED** [review]

**Comment by sv641** [comment]
After this commit is done, we should seriously consider maintaining a BUILDROOT.
move tests build output to BUILDROOT.

File: .github/workflows/km-ci-workflow.yaml [hunk 2]
```diff
@@ -141,56 +142,107 @@ jobs:
...
-        run: |
-          make -C payloads -j runenv-image
-          make -C payloads -j push-demo-runenv-image
-
-      # Note: custom Python build needs to happen after python payload build
-      - name: Python.km custom build
-        run: |
-          make -C payloads/python build-modules pack-modules CUSTOM_MODULES="markupsafe wrapt"
-          make -C payloads/python custom CUSTOM_MODULES="markupsafe wrapt"
+      # - name: Build payloads and create test images
```
[the remark this answers is not in the document]
Comment (ez-leka) [reply .github/workflows/km-ci-workflow.yaml:145]: Sorry, bad habit of commenting staff out. It did

File: .github/workflows/km-ci-workflow.yaml [hunk 3]
```diff
@@ -141,56 +142,107 @@ jobs:
...
           name: km-build
           path: |
             build/cloud/k8s/deploy/shim
-            build/km/km
-            tests/hello_test.km
-            build/container-runtime/krun-label-trigger
+            build/opt/kontain/bin
+            build/opt/kontain/lib
+            build/opt/kontain/include
+            tests/hello_test.kmd
```
[the remark this answers is not in the document]
Comment (ez-leka) [reply .github/workflows/km-ci-workflow.yaml:174]: we always used it, it was just pre-build in tests and there. Now that payloads are built as part of separate job, we have to pass it as part of artifact

**Review by johnmuth81 — APPROVED** [review]
Checked out and ran on my local system after removing /opt/kontain. Seems to work fine.

**Comment by ez-leka** [comment]
Thank you!

Elana

File: .github/workflows/km-ci-workflow.yaml [hunk 4]
```diff
@@ -141,56 +142,89 @@ jobs:
...
-            /opt/kontain/bin/shim
+            build/opt/kontain/bin/km
+            build/opt/kontain/bin/kontain-gcc
+            build/opt/kontain/bin/krun
+            build/opt/kontain/bin/shim
           retention-days: 7
 
+      # passign hello_test.kmd that is required by testenv of payloads - it already has correct RPATH.
+      # passing hello_test.c and libhelper.a
+      # because we will need to rebuid hello_test.kmd with correct RPATH for runtime environments
```
Comment (gnode1) [inline .github/workflows/km-ci-workflow.yaml:166]: This line and two above sounds contradictory, or at least confusing

File: Makefile [hunk 5]
```diff
@@ -44,26 +44,57 @@ clang-format-check:
...
+${KM_BIN_RELEASE}: ${KM_RELEASE} ## Build a release tar.gz file for KM runtime binaries
+	tar -czvf $@ ${KM_BINARIES}
+
+clean-release: ## Clean the release tar files
+	rm -f ${KM_RELEASE} ${KM_BIN_RELEASE}
+
+publish-release: ## Publish release with RELEASE_TAG to github
+	cd ${TOP}/tools/release; ./release_km.py ${KM_RELEASE} ${KM_BIN_RELEASE} ${KM_KKM_RELEASE} --version ${RELEASE_TAG} --message "${RELEASE_MESSAGE}"
+
+
```
Comment (gnode1) [inline Makefile:88]: extra empty line

File: make/images.mk [hunk 6]
```diff
@@ -73,11 +73,36 @@ TESTENV_PATH ?= .
...
+						opt/kontain/alpine-lib/usr/lib/libffi.so \
+						opt/kontain/alpine-lib/usr/lib/libffi.so.6 \
+						opt/kontain/alpine-lib/usr/lib/libffi.so.7 \
+						opt/kontain/alpine-lib/usr/lib/libffi.so.7.1.0 \
+						opt/kontain/alpine-lib/usr/lib/libgcc_s.so \
+						opt/kontain/runtime/libpthread.so
+	$(if ${TESTENV_EXTRA_FILES},cp -r --preserve=links ${TESTENV_EXTRA_FILES} ${TESTENV_PATH})
+endef
+
+
```
Comment (gnode1) [inline make/images.mk:99]: empty line

File: make/images.mk [hunk 7]
```diff
@@ -73,11 +73,36 @@ TESTENV_PATH ?= .
...
+testenv_cleanup = rm ${TESTENV_PATH}/extras.tar.gz
+testenv_cleanup_extras=$(if ${TESTENV_EXTRA_FILES},rm  ${TESTENV_PATH}/${TESTENV_EXTRA_FILES} )
+
+## build test image with test tools and code
+## requires building
```
Comment (gnode1) [inline make/images.mk:104]: what does this mean?

File: make/locations.mk [hunk 8]
```diff
@@ -56,16 +54,24 @@ BLDDIR := $(abspath ${BLDTOP}/${FROMTOP}/${BLDTYPE})
...
 KM_OPT_BIN := ${KM_OPT}/bin
 KM_OPT_INC := ${KM_OPT}/include
 KM_OPT_LIB := ${KM_OPT}/lib
 KM_OPT_RT := ${KM_OPT}/runtime
 KM_OPT_ALPINELIB := ${KM_OPT}/alpine-lib
-KM_LDSO := ${BLDTOP}/runtime/libc.so
+KM_LDSO := ${KM_OPT_RT}/libc.so
+KM_BIN := ${KM_OPT_BIN}/km
+
+
```
Comment (gnode1) [inline make/locations.mk:73]: empty line

**Review by gnode1 — APPROVED** [review]
LGTM with nits and questions

File: .github/workflows/km-ci-workflow.yaml (hunk 4 above)
Comment (ez-leka) [reply .github/workflows/km-ci-workflow.yaml:166]: Not sure what is confusing - using already existing .kmd for testing and will re-build for runner/demo-runenv

File: km-releases/kontain-install.sh [hunk 9]
```diff
@@ -27,21 +29,25 @@ set -e; [ "$TRACE" ] && set -x
 source /etc/os-release
 [ "$ID" != "fedora" -a "$ID" != "ubuntu" -a "$ID" != "amzn" ] && echo "Unsupported linux distribution: $ID" && exit 1
 
+
```
[the remark this answers is not in the document]
Comment (ez-leka) [reply km-releases/kontain-install.sh:32]: done

File: tests/km_core_tests.bats [hunk 10]
```diff
@@ -27,7 +27,7 @@ load test_helper
 
 # not_needed_{generic,static,dynamic,shared} - skip since it's not needed
 # todo_{generic,static,dynamic,shared} - skip since it's a TODO
-not_needed_generic=
+not_needed_generic=perf
```
[the remark this answers is not in the document]
Comment (ez-leka) [reply tests/km_core_tests.bats:30]: no idea - I think you did it when we were fighting staff out. I'll remove perf?

**Pull request merged**